src/generic/C1_helper.h

```diff
@@ -0,0 +1,1092 @@
+ 
+
+//==============================================================================
+/// A function that upgrades straight sided elements to be curved. This involves
+/// Setting up the parametric boundary, F(s) and the first derivative F'(s)
+/// We also need to set the edge number of the upgraded element and the positions
+/// of the nodes j and k (defined below) and set which edge (k) is to be exterior
+///            @ k               
+///           /(                 
+///          /. \                
+///         /._._)               
+///      i @     @ j             
+/// For RESTING or FREE boundaries we need to have a C2 CONTINUOUS boundary
+/// representation. That is we need to have a continuous 2nd derivative defined
+/// too. This is well discussed in by [Zenisek 1981] (Aplikace matematiky ,
+/// Vol. 26 (1981), No. 2, 121--141). This results in the necessity for F''(s)
+/// as well.
+/// The smooth mesh/domain boundaries are available via the map c1_curviline_pt.
+/// hierher should we create/store/delete the c1_curviline_pt things in the
+/// triangle mesh (either by default or when required). would probaly fix that
+/// leak problem.
+//=============================================================================
+ extern void upgrade_edge_elements_to_curved_boundaries(
+  Mesh* bulk_mesh_pt, 
+  std::map<unsigned,C1CurviLine*> c1_curviline_pt);
+
+
+ 
+//======================================================================
+/// Function to set up rotated nodes on the boundaries of the triangle
+/// mesh pointed to by bulk_mesh_pt. Necessary if we want to set
+/// up physical boundary conditions on a curved boundary with Hermite type dofs.
+/// For example if we know w(n,t) = f(t) (where n and t are the
+/// normal and tangent to a boundary) we ALSO know dw/dt and d2w/dt2.
+/// NB no rotation is needed if the edges are completely free!
+/// The smooth mesh/domain boundaries are available via the map c1_curviline_pt.
+//======================================================================
+ extern void rotate_edge_degrees_of_freedom(
+  Mesh* bulk_mesh_pt, 
+  std::map<unsigned,C1CurviLine*> c1_curviline_pt);
+
+ 
+ 
+//======================================================================
+/// Helper function to upgrade a triangle mesh to C1 computations.
+/// -- split elements that have multiple edges on a boundary (so the number
+///    of elements in the mesh changes!
+/// -- duplicate corner nodes (so the number of nodes in the mesh changes
+/// -- create DuplicateNodeConstraintElements that constrain the degrees of freedom
+///    at the newly created nodes to ensure continuity of the displacements
+///    These elements are added to the mesh pointed to by constraint_mesh_pt
+/// -- rotate the degrees of freedom on the curvilinear boundaries so they
+///    represent displacements and derivatives in the tangential and
+///    normal direction // hierher Aidan true?
```

Comment:
Ambiguously worded here. The displacements are still cartesian oriented, just the coordinates are "rotated":
(u_x, u_y, u_z)(n, t)

Reply:
done